### DISCUSSION TITLE
Water_flow based on DeepXDE

### DISCUSSION THREAD

**Discussion by xmuyao11** [lululxvi/deepxde#1333] [Q&A]
Hi,Lulu.During my research on fluid based on PINN,I find some queetions.I can't understand why I can't get a good result when use water instead of air.My project is to simulate the fluid flow inside the tube.Could you please help me,please?

**Comment by lululxvi** [comment]
More details are required.

**Reply by xmuyao11** [reply]
Thanks for your reply.Here are my losses during my research.
![befa3b97cd8d1e4f2a10ce344e51ae6](https://github.com/lululxvi/deepxde/assets/134580477/dbe84c54-0b10-40bf-beb6-f0dd941f043e)

This is air flowing in a circular tube.Its density is 1kg/m3 and its viscosity is 0.001.During this calculation,I can get a good result.
![37de46fbcbf0919e713dddd84405faf](https://github.com/lululxvi/deepxde/assets/134580477/1abbee9a-88ff-42b7-bb6c-fef48bd137e8)

However,when I use water whose density is 1000 instead of air,I can't get a good result,my loss won't decrease,and I don't change other options.

Please give me some suggestions.

**Reply by ryabs07** [reply]
Maybe it is a problem of scale? By scaling/non-dimensionalizing the PDEs, you may achieve better convergence and more stable training for your PINN model. It will help making the problem independent of the specific fluid properties and system dimensions.

**Reply by david-e-gomes** [reply]
### 1 - I have the same problem...
 Im trying to solve an inverse problem with NS equations but for an air flow with density of 1 a viscosity of 0,01 it works, when I try to do with density of 1000 and viscosity of 0,001 for water the predicted parameters doesn't fit with CFD reference (Obviously I changed the supervised data)... I supose that the problem is a problem about the magnitude of parameters probably but I don't know how to solve it... this are pde equations for air flow:

```python
# Definir equações de Navier-Stokes para o problema inverso
def Navier_Stokes_Equation(x, y):
    # Definir densidade e viscosidade da água (em unidades SI)
    rho = 1  # Densidade (adimensional no contexto)
    mu = 0.01  # Viscosidade dinâmica (Pa·s ou Ns/m^2)

    u = y[:, 0:1]  # Velocidade na direção x
    v = y[:, 1:2]  # Velocidade na direção y
    p = y[:, 2:3]  # Pressão

    # Derivadas de primeira ordem
    du_x = dde.grad.jacobian(y, x, i=0, j=0)
    du_y = dde.grad.jacobian(y, x, i=0, j=1)
    dv_x = dde.grad.jacobian(y, x, i=1, j=0)
    dv_y = dde.grad.jacobian(y, x, i=1, j=1)
    dp_x = dde.grad.jacobian(y, x, i=2, j=0)
    dp_y = dde.grad.jacobian(y, x, i=2, j=1)

    # Derivadas de segunda ordem (laplaciano)
    du_xx = dde.grad.hessian(y, x, component=0, i=0, j=0)
    du_yy = dde.grad.hessian(y, x, component=0, i=1, j=1)
    dv_xx = dde.grad.hessian(y, x, component=1, i=0, j=0)
    dv_yy = dde.grad.hessian(y, x, component=1, i=1, j=1)

    # Continuidade e equações de momento usando valores conhecidos de densidade e viscosidade
    continuity = du_x + dv_y
    x_momentum = rho * (u * du_x + v * du_y) + dp_x - mu * (du_xx + du_yy)
    y_momentum = rho * (u * dv_x + v * dv_y) + dp_y - mu * (dv_xx + dv_yy)

    return [continuity, x_momentum, y_momentum]

# Exemplo de uso para a zona 'interior'
[x_train_interior, y_train_interior, u_train_interior, v_train_interior, p_train_interior] = load_training_data_from_zones(num_points_per_zone=1325, selected_zone='interior-surface_body')
[x_train_inlet, y_train_inlet, u_train_inlet, v_train_inlet, p_train_inlet] = load_training_data_from_zones(num_points_per_zone=9, selected_zone='inlet')
[x_train_outlet, y_train_outlet, u_train_outlet, v_train_outlet, p_train_outlet] = load_training_data_from_zones(num_points_per_zone=14, selected_zone='outlet')
[x_train_walls, y_train_walls, u_train_walls, v_train_walls, p_train_walls] = load_training_data_from_zones(num_points_per_zone=176, selected_zone='walls')

# Separar os dados do interior
interior_coordinates = np.column_stack((x_train_interior, y_train_interior))
interior_u = u_train_interior
interior_v = v_train_interior

# Separar as boundary conditions
boundary_coordinates = np.vstack((
    np.column_stack((x_train_inlet, y_train_inlet)),
    np.column_stack((x_train_outlet, y_train_outlet)),
    np.column_stack((x_train_walls, y_train_walls))
))

# Combinar todas as coordenadas (interior e fronteiras) para observação
all_x_coords = np.concatenate([x_train_interior, x_train_inlet, x_train_outlet, x_train_walls], axis=0)
all_y_coords = np.concatenate([y_train_interior, y_train_inlet, y_train_outlet, y_train_walls], axis=0)
all_u = np.concatenate([u_train_interior, u_train_inlet, u_train_outlet, u_train_walls], axis=0)
all_v = np.concatenate([v_train_interior, v_train_inlet, v_train_outlet, v_train_walls], axis=0)

# Criar matriz de coordenadas combinadas para observação
all_xy = np.hstack((all_x_coords, all_y_coords))

# Definir a geometria a partir das coordenadas do domínio
geom = dde.geometry.PointCloud(interior_coordinates, boundary_points=boundary_coordinates)

# Definir os observadores combinados para u e v
observe_u = dde.icbc.PointSetBC(all_xy, all_u, component=0)
observe_v = dde.icbc.PointSetBC(all_xy, all_v, component=1)

# Definir os dados para o modelo PINN
data = dde.data.PDE(
    geom,
    Navier_Stokes_Equation,
    [observe_u, observe_v],  # Observadores separados
    num_domain=1325,  # Pontos internos (collocation points)
    num_boundary=200,
    anchors=all_xy  # Combinação total para âncoras
)

# Configurar a rede neural (PINN)
layer_size = [2] + [40] * 8 + [3]  # 2 entradas (x, y) e 3 saídas (u, v, p)
activation = "tanh"
initializer = "Glorot uniform"
net = dde.maps.FNN(layer_size, activation, initializer)

# Criar o modelo PINN
model = dde.Model(data, net)

# Configurar early stopping
early_stopping = dde.callbacks.EarlyStopping(min_delta=1e-6, patience=5000)

# Compilar o modelo e treinar com early stopping
model.compile("adam", lr=1e-3)
loss_history, train_state = model.train(iterations=30000, callbacks=[early_stopping], display_every=1000, disregard_previous_best=True)

# Salvar e plotar resultados
dde.saveplot(loss_history, train_state, issave=True, isplot=True)
``` 

With this code I got very good results for laminar flows, very good reconstruction: 
![image](https://github.com/user-attachments/assets/25e89e8b-afc2-4763-a7ba-f1adc14c61fd)
![image](https://github.com/user-attachments/assets/1418d157-38da-49fa-954a-fd901df01ca5)
![image](https://github.com/user-attachments/assets/065a8218-f070-4849-b2dc-560f7016a22f)


However just changing the density for 1000 and viscosity for 0,001 (and changing the data reference for water simulation) the results are like this... 
![image](https://github.com/user-attachments/assets/ade95b88-608b-41cb-92ff-3b4c7eeb0ee3)
![image](https://github.com/user-attachments/assets/e2075143-ce84-47cc-9020-ddf6fe8630d0)
![image](https://github.com/user-attachments/assets/ac9ceea3-b836-40f9-b178-83da6919e280)

There are other form to write NS equations for water that fits well with deepxde library?


### 2- Other problem is about normalization.. 
I know that for this airflow case I don't need normalization but when I apply it within a range of [0,1] for parameters u, v and p, the results are not good... This is important after for my case with water that the velocities and pressure are from very different orders of magnitude

### 3 -Is there any trend about turbloence modeling in PINNs?

Thank you in advance @lululxvi @xmuyao11 @ryabs07
